src/generators/markdowndocgenerator/index.ts

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { IDocumentWidget, MimeDocument } from '@jupyterlab/docregistry';
+
+import { FileEditor, IEditorTracker } from '@jupyterlab/fileeditor';
+
+import { TableOfContentsRegistry } from '../../registry';
+
+import { SharedMethods, INotebookHeading } from '../shared';
+
+import { IInstanceTracker, ISanitizer } from '@jupyterlab/apputils';
+
+import { MarkdownDocGeneratorOptionsManager } from './optionsmanager';
+
+import { TableOfContents } from '../../toc';
+
+import { markdownDocItemRenderer } from './itemrenderer';
+
+import { markdownDocGeneratorToolbar } from './toolbargenerator';
+
+/**
+ * Create a TOC generator for markdown files.
+ *
+ * @param tracker: A file editor tracker.
+ *
+ * @returns A TOC generator that can parse markdown files.
+ */
+export function createMarkdownGenerator(
+  tracker: IEditorTracker,
+  widget: TableOfContents
+): TableOfContentsRegistry.IGenerator<IDocumentWidget<FileEditor>> {
+  const options = new MarkdownDocGeneratorOptionsManager(widget, {
+    needNumbering: true
+  });
+  return {
+    tracker,
+    usesLatex: true,
+    options: options,
+    toolbarGenerator: () => {
```

Comment:
I see you just changed this. I was actually starting to think you were right the first time. All the other functions that are provided have the signature `widget => something`. Maybe this should as well. It would then be a little easier to tailor the toolbar to the current widget as well, since it will get a reference to it upon toolbar creation. Maybe that cleaner API would be worth the performance hit of recreating the toolbar.

What do you think @zuoyuanh?

Reply:
Yeah I'm sorry I can't figure out a way to store the toolbar without making the API more complicated, but I feel like the old API is easier for user to add their own generator. Actually with the new API user can still use the old generators without any changes, but I feel like other people may be confused when they trying to look at the existing API.

I can remove the commit from master branch that made this change and keep it in another branch in case we need to add it back later or for reference if we want to create a better way to store the toolbar in the future.

Thank you!

Reply:
Sounds good. Let's go with `widget => toolbar` for the function signature, then.